c++/src/capnp/compat/json.c++

```diff
@@ -372,6 +382,251 @@ Orphan<DynamicValue> JsonCodec::decode(
+    output.setString(consumeQuotedString());
+  }
+
+  void parseArray(JsonValue::Builder &output) {
+    // TODO(perf): Using orphans leaves holes in the message. It's expected
+    // that a JsonValue is used for interop, and won't be sent or written as a
+    // Cap'n Proto message.
+    kj::Vector<Orphan<JsonValue>> values;
+    auto orphanage = Orphanage::getForMessageContaining(output);
+
+    consume('[');
+    while (consumeWhitespace(), nextChar() != ']') {
+      auto orphan = orphanage.newOrphan<JsonValue>();
+      auto builder = orphan.get();
+      parseValue(builder);
+      values.add(kj::mv(orphan));
+
+      if (consumeWhitespace(), nextChar() != ']') {
+        // TODO(soon): This incorrectly allows a trailing comma.
+        consume(',');
+      }
+    }
+
+    output.initArray(values.size());
+    auto array = output.getArray();
+
+    for (auto i : kj::indices(values)) {
+      array.adoptWithCaveats(i, kj::mv(values[i]));
+    }
+
+    consume(']');
+  }
+
+  void parseObject(JsonValue::Builder &output) {
+    kj::Vector<Orphan<JsonValue::Field>> fields;
+    auto orphanage = Orphanage::getForMessageContaining(output);
+
+    consume('{');
+    while (consumeWhitespace(), nextChar() != '}') {
+      auto orphan = orphanage.newOrphan<JsonValue::Field>();
+      auto builder = orphan.get();
+
+      builder.setName(consumeQuotedString());
+
+      consumeWhitespace();
+      consume(':');
+      consumeWhitespace();
+
+      auto valueBuilder = builder.getValue();
+      parseValue(valueBuilder);
+
+      fields.add(kj::mv(orphan));
+
+      if (consumeWhitespace(), nextChar() != '}') {
+        // TODO(soon): This incorrectly allows a trailing comma.
+        consume(',');
+      }
+    }
+
+    output.initObject(fields.size());
+    auto object = output.getObject();
+
+    for (auto i : kj::indices(fields)) {
+      object.adoptWithCaveats(i, kj::mv(fields[i]));
+    }
+
+    consume('}');
+  }
+
+  char nextChar() {
+    return remaining_.front();
+  }
+
+  void advance(size_t numBytes = 1) {
+    KJ_REQUIRE(numBytes <= remaining_.size(), "JSON message ends prematurely.");
+    remaining_ = kj::arrayPtr(remaining_.begin() + numBytes, remaining_.end());
+  }
+
+  void advanceTo(const char *newPos) {
+    KJ_REQUIRE(remaining_.begin() <= newPos && newPos < remaining_.end(),
+        "JSON message ends prematurely.");
+    remaining_ = kj::arrayPtr(newPos, remaining_.end());
+  }
+
+  void consume(char expected) {
+    char current = nextChar();
+    KJ_REQUIRE(current == expected, "Unexpected character in JSON message.");
+
+    advance();
+  }
+
+  void consume(kj::ArrayPtr<const char> expected) {
+    KJ_REQUIRE(remaining_.size() >= expected.size());
+
+    auto prefix = remaining_.slice(0, expected.size());
+    KJ_REQUIRE(prefix == expected, "Unexpected input in JSON message.");
+
+    advance(expected.size());
+  }
+
+  kj::ArrayPtr<const char> consumeWhile(kj::Function<bool(char)> predicate) {
```

Comment:
kj::Function adds virtual call overhead which could be expensive when calling for every byte. Instead, make this a template:

    template <typename Predicate>
    kj::ArrayPtr<const char> consumeWhile(Predicate&& predicate) {